{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### params \n",
    "# get the option of pi beta and alle freq chosen \n",
    "chr_number = 5\n",
    "n_ecotypes = 225 \n",
    "vcf_file = '../data/chr5_grenenet.vcf'\n",
    "\n",
    "#pi_option =  snakemake.params['pi']\n",
    "#beta = snakemake.params['beta']\n",
    "#print(beta)\n",
    "#alelle_freq_option = snakemake.params['allele_freq']\n",
    "\n",
    "#get the actual values\n",
    "n_optima = 15\n",
    "\n",
    "pi = 0.000005\n",
    "\n",
    "allele_freq = [0.05, 0.5]\n",
    "lower_bound = float(allele_freq[0])\n",
    "upper_bound = float(allele_freq[1])\n",
    "\n",
    "#bed_sc = snakemake.output[\"bed_sc\"]\n",
    "#optima_slim = snakemake.output[\"slim\"]\n",
    "######\n",
    "\n",
    "## get vcf file\n",
    "vcf_og = allel.read_vcf(vcf_file)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scale(beta_values):\n",
    "    min_value = np.min(beta_values)\n",
    "    max_value = np.max(beta_values)\n",
    "    beta_values = ((beta_values - min_value) / (max_value - min_value)) * (5 - 1) + 1\n",
    "    return beta_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def getting_selection_coef(vcf_og, pi, beta, lower_bound, upper_bound):\n",
    "    ### calculation of the number of contributing positions and effect sizes of each ###################\n",
    "    pos = vcf_og['variants/POS']\n",
    "    ## get all the pos from the vcf file\n",
    "    n_pos = len(pos)\n",
    "    geno_og = vcf_og[\"calldata/GT\"]\n",
    "    ## freq of alternative alleles\n",
    "    alt_al_count = geno_og.sum(axis=2).sum(axis=1)\n",
    "    n_pos_contr = round(pi*n_pos)\n",
    "    \n",
    "    alelle_dist = pd.DataFrame({'alt_al_count':alt_al_count, 'pos':pos})\n",
    "\n",
    "    alelle_dist['alt_al_freq'] = alelle_dist['alt_al_count'] / (n_ecotypes*2)\n",
    "\n",
    "    if beta == 'betaprop': \n",
    "        alelle_dist = alelle_dist.copy()\n",
    "        beta_values = alelle_dist.loc[:,('alt_al_freq')] * 5\n",
    "        beta_values = scale(beta_values)\n",
    "        alelle_dist.loc[:,'beta'] = beta_values\n",
    "\n",
    "    elif beta == 'betanonprop':\n",
    "        alelle_dist = alelle_dist.copy()\n",
    "        beta_values = 5 - (alelle_dist.loc[:,('alt_al_freq')] * 5 )\n",
    "        beta_values = scale(beta_values)\n",
    "        alelle_dist.loc[:,'beta'] = beta_values\n",
    "\n",
    "    alelle_dist = alelle_dist[(alelle_dist['alt_al_freq'] > lower_bound) & (alelle_dist['alt_al_freq'] <= upper_bound)]\n",
    "\n",
    "    pos_contr = alelle_dist['pos'].sample(n_pos_contr)\n",
    "\n",
    "    alelle_dist = alelle_dist[alelle_dist['pos'].isin(pos_contr)]\n",
    "\n",
    "    effect_sizes = [np.random.normal(0, beta, 1) for beta in alelle_dist['beta']]\n",
    "    effect_sizes = effect_sizes.copy()\n",
    "\n",
    "    alelle_dist.loc[:,'effect_sizes'] = np.ravel(effect_sizes).copy()\n",
    "    alelle_dist.loc[:,'effect_sizes'] = alelle_dist.loc[:,'effect_sizes'].round(4)\n",
    "    effect_sizes = np.round(alelle_dist['effect_sizes'], 4)\n",
    "\n",
    "    pos = pd.Series(pos)\n",
    "    pos.name = 'pos'\n",
    "    bed_file = pd.merge(pos, alelle_dist[['pos', 'effect_sizes']], on='pos', how='outer').fillna(0)\n",
    "    bed_file['positions_from'] = bed_file['pos'] - 1\n",
    "    bed_file['chr'] = chr_number\n",
    "    bed_file = bed_file[['chr', 'positions_from', 'pos', 'effect_sizes']]\n",
    "\n",
    "    return bed_file\n",
    "\n",
    "def calc_optima(bed_file, vcf_og):\n",
    "    geno_og = vcf_og[\"calldata/GT\"]\n",
    "    alt_al_per_pos = geno_og.sum(axis=2) #.sum(axis=1)\n",
    "    phenotypes = []\n",
    "\n",
    "    for i in range(alt_al_per_pos.shape[1]):\n",
    "        gen_effectsize = np.multiply(alt_al_per_pos[:, i] , np.array(bed_file['effect_sizes']))\n",
    "        phenotypes.append(gen_effectsize.sum())\n",
    "\n",
    "    max_pheno = max(phenotypes)\n",
    "\n",
    "    min_pheno = min(phenotypes)\n",
    "\n",
    "    optima = np.random.uniform(min_pheno, max_pheno, n_optima - 2)\n",
    "    optima = np.append(optima, [min_pheno, max_pheno])\n",
    "    return optima\n",
    "\n",
    "bed_file = getting_selection_coef(vcf_og, pi, beta, lower_bound, upper_bound)\n",
    "bed_file.to_csv(bed_sc, sep='\\t', header=False, index=False) # selection_coef_chr5.bed\n",
    "\n",
    "pheno_optima = calc_optima(bed_file, vcf_og)\n",
    "pd.Series(pheno_optima).to_csv(optima_slim, header=None, sep='\\n', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "beta = 'betanonprop'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "### calculation of the number of contributing positions and effect sizes of each ###################\n",
    "pos = vcf_og['variants/POS']\n",
    "## get all the pos from the vcf file\n",
    "n_pos = len(pos)\n",
    "geno_og = vcf_og[\"calldata/GT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        ...,\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0]],\n",
       "\n",
       "       [[0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        ...,\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0]],\n",
       "\n",
       "       [[0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        ...,\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0]],\n",
       "\n",
       "       ...,\n",
       "\n",
       "       [[0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        ...,\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0]],\n",
       "\n",
       "       [[0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        ...,\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0]],\n",
       "\n",
       "       [[0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        ...,\n",
       "        [0, 0],\n",
       "        [0, 0],\n",
       "        [0, 0]]], dtype=int8)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geno_og"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## freq of alternative alleles\n",
    "alt_al_count = geno_og.sum(axis=2).sum(axis=1)\n",
    "n_pos_contr = round(0.1*n_pos)\n",
    "\n",
    "alelle_dist = pd.DataFrame({'alt_al_count':alt_al_count, 'pos':pos})\n",
    "\n",
    "alelle_dist['alt_al_freq'] = alelle_dist['alt_al_count'] / (n_ecotypes*2)\n",
    "\n",
    "if beta == 'betaprop': \n",
    "    alelle_dist = alelle_dist.copy()\n",
    "    beta_values = alelle_dist.loc[:,('alt_al_freq')] * 5\n",
    "    beta_values = scale(beta_values)\n",
    "    alelle_dist.loc[:,'beta'] = beta_values\n",
    "\n",
    "elif beta == 'betanonprop':\n",
    "    alelle_dist = alelle_dist.copy()\n",
    "    beta_values = 5 - (alelle_dist.loc[:,('alt_al_freq')] * 5 )\n",
    "    beta_values = scale(beta_values)\n",
    "    alelle_dist.loc[:,'beta'] = beta_values\n",
    "\n",
    "alelle_dist = alelle_dist[(alelle_dist['alt_al_freq'] > lower_bound) & (alelle_dist['alt_al_freq'] <= upper_bound)]\n",
    "\n",
    "pos_contr = alelle_dist['pos'].sample(n_pos_contr)\n",
    "\n",
    "alelle_dist = alelle_dist[alelle_dist['pos'].isin(pos_contr)]\n",
    "\n",
    "effect_sizes = [np.random.normal(0, beta, 1) for beta in alelle_dist['beta']]\n",
    "effect_sizes = effect_sizes.copy()\n",
    "\n",
    "alelle_dist.loc[:,'effect_sizes'] = np.ravel(effect_sizes).copy()\n",
    "alelle_dist.loc[:,'effect_sizes'] = alelle_dist.loc[:,'effect_sizes'].round(4)\n",
    "effect_sizes = np.round(alelle_dist['effect_sizes'], 4)\n",
    "\n",
    "pos = pd.Series(pos)\n",
    "pos.name = 'pos'\n",
    "bed_file = pd.merge(pos, alelle_dist[['pos', 'effect_sizes']], on='pos', how='outer').fillna(0)\n",
    "bed_file['positions_from'] = bed_file['pos'] - 1\n",
    "bed_file['chr'] = chr_number\n",
    "bed_file = bed_file[['chr', 'positions_from', 'pos', 'effect_sizes']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alt_al_count</th>\n",
       "      <th>pos</th>\n",
       "      <th>alt_al_freq</th>\n",
       "      <th>beta</th>\n",
       "      <th>effect_sizes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32</td>\n",
       "      <td>620146</td>\n",
       "      <td>0.071111</td>\n",
       "      <td>4.962264</td>\n",
       "      <td>-2.1544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>80</td>\n",
       "      <td>5415374</td>\n",
       "      <td>0.177778</td>\n",
       "      <td>4.905660</td>\n",
       "      <td>8.3796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>128</td>\n",
       "      <td>5571785</td>\n",
       "      <td>0.284444</td>\n",
       "      <td>4.849057</td>\n",
       "      <td>-0.4071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>32</td>\n",
       "      <td>6346978</td>\n",
       "      <td>0.071111</td>\n",
       "      <td>4.962264</td>\n",
       "      <td>6.3330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>64</td>\n",
       "      <td>6733596</td>\n",
       "      <td>0.142222</td>\n",
       "      <td>4.924528</td>\n",
       "      <td>-7.7236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>192</td>\n",
       "      <td>10015566</td>\n",
       "      <td>0.426667</td>\n",
       "      <td>4.773585</td>\n",
       "      <td>5.2770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>80</td>\n",
       "      <td>12193845</td>\n",
       "      <td>0.177778</td>\n",
       "      <td>4.905660</td>\n",
       "      <td>1.7751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>200</td>\n",
       "      <td>12971731</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>4.764151</td>\n",
       "      <td>-7.6727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>32</td>\n",
       "      <td>14211701</td>\n",
       "      <td>0.071111</td>\n",
       "      <td>4.962264</td>\n",
       "      <td>-3.6997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>208</td>\n",
       "      <td>15927575</td>\n",
       "      <td>0.462222</td>\n",
       "      <td>4.754717</td>\n",
       "      <td>0.7993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>80</td>\n",
       "      <td>16388039</td>\n",
       "      <td>0.177778</td>\n",
       "      <td>4.905660</td>\n",
       "      <td>1.8032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>80</td>\n",
       "      <td>19224891</td>\n",
       "      <td>0.177778</td>\n",
       "      <td>4.905660</td>\n",
       "      <td>-6.1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>32</td>\n",
       "      <td>19898795</td>\n",
       "      <td>0.071111</td>\n",
       "      <td>4.962264</td>\n",
       "      <td>7.5341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>96</td>\n",
       "      <td>20307611</td>\n",
       "      <td>0.213333</td>\n",
       "      <td>4.886792</td>\n",
       "      <td>-5.5198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>48</td>\n",
       "      <td>22401709</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>4.943396</td>\n",
       "      <td>11.4958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>32</td>\n",
       "      <td>22940288</td>\n",
       "      <td>0.071111</td>\n",
       "      <td>4.962264</td>\n",
       "      <td>-4.2216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>136</td>\n",
       "      <td>25602568</td>\n",
       "      <td>0.302222</td>\n",
       "      <td>4.839623</td>\n",
       "      <td>-2.5924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>48</td>\n",
       "      <td>26413869</td>\n",
       "      <td>0.106667</td>\n",
       "      <td>4.943396</td>\n",
       "      <td>-3.6470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>192</td>\n",
       "      <td>26533634</td>\n",
       "      <td>0.426667</td>\n",
       "      <td>4.773585</td>\n",
       "      <td>-8.4994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>64</td>\n",
       "      <td>26533729</td>\n",
       "      <td>0.142222</td>\n",
       "      <td>4.924528</td>\n",
       "      <td>2.1708</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     alt_al_count       pos  alt_al_freq      beta  effect_sizes\n",
       "4              32    620146     0.071111  4.962264       -2.1544\n",
       "27             80   5415374     0.177778  4.905660        8.3796\n",
       "29            128   5571785     0.284444  4.849057       -0.4071\n",
       "33             32   6346978     0.071111  4.962264        6.3330\n",
       "34             64   6733596     0.142222  4.924528       -7.7236\n",
       "56            192  10015566     0.426667  4.773585        5.2770\n",
       "70             80  12193845     0.177778  4.905660        1.7751\n",
       "76            200  12971731     0.444444  4.764151       -7.6727\n",
       "86             32  14211701     0.071111  4.962264       -3.6997\n",
       "90            208  15927575     0.462222  4.754717        0.7993\n",
       "93             80  16388039     0.177778  4.905660        1.8032\n",
       "119            80  19224891     0.177778  4.905660       -6.1280\n",
       "124            32  19898795     0.071111  4.962264        7.5341\n",
       "127            96  20307611     0.213333  4.886792       -5.5198\n",
       "150            48  22401709     0.106667  4.943396       11.4958\n",
       "153            32  22940288     0.071111  4.962264       -4.2216\n",
       "167           136  25602568     0.302222  4.839623       -2.5924\n",
       "172            48  26413869     0.106667  4.943396       -3.6470\n",
       "174           192  26533634     0.426667  4.773585       -8.4994\n",
       "175            64  26533729     0.142222  4.924528        2.1708"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alelle_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4      -2.1544\n",
       "27      8.3796\n",
       "29     -0.4071\n",
       "33      6.3330\n",
       "34     -7.7236\n",
       "56      5.2770\n",
       "70      1.7751\n",
       "76     -7.6727\n",
       "86     -3.6997\n",
       "90      0.7993\n",
       "93      1.8032\n",
       "119    -6.1280\n",
       "124     7.5341\n",
       "127    -5.5198\n",
       "150    11.4958\n",
       "153    -4.2216\n",
       "167    -2.5924\n",
       "172    -3.6470\n",
       "174    -8.4994\n",
       "175     2.1708\n",
       "Name: effect_sizes, dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([2.07706234]),\n",
       " array([4.18162726]),\n",
       " array([4.16041672]),\n",
       " array([0.67296217]),\n",
       " array([5.20508181]),\n",
       " array([1.0418717]),\n",
       " array([4.11958659]),\n",
       " array([0.06750712]),\n",
       " array([1.33297966]),\n",
       " array([2.79536142]),\n",
       " array([1.17411296]),\n",
       " array([2.08406651]),\n",
       " array([3.04128033]),\n",
       " array([3.05312375]),\n",
       " array([3.90863213]),\n",
       " array([5.07889731]),\n",
       " array([2.4624949]),\n",
       " array([2.95655842]),\n",
       " array([7.43460793]),\n",
       " array([0.59851202])]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[abs(np.random.normal(0, beta, 1)) for beta in alelle_dist['beta']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pipeline_slim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
